### PR TITLE
Parallel states

### PR DESCRIPTION
**Pull request by thatch45**
YES! You heard that right PARALLEL STATES!!!!!!

This PR allows you to set `parallel: True` on a state and it runs in a separate process!

Salt States just went from fast to ludicrous speed!

I am having a hard time figuring out a good way to test these to make sure states are running in parallel in an automated way, so I am open to suggestions.

### CODE REVIEW

**Comment by gtmanfred** [comment]
That was fast.

File: salt/state.py [hunk 1]
```diff
@@ -1920,6 +1983,35 @@ def check_failhard(self, low, running):
...
+                    except OSError, IOError:
+                        ret = {'result': False,
+                               'comment': 'Parallel cache failure',
+                               'name': running[tag]['name'],
+                               'changes': {}}
+                    running[tag].update(ret)
+                    running[tag].pop('proc')
+                else:
+                    retset.add(False)
+        return False if False in retset else True
```
Comment (KaiSforza) [inline salt/state.py]: Why not use `return False not in retset`? Just think it makes it a bit more readable.

File: salt/state.py [hunk 2]
```diff
@@ -1645,6 +1649,56 @@ def requisite_in(self, high):
...
+        if not os.path.isdir(troot):
+            os.makedirs(troot)
+        with open(tfile, 'wb+') as fp_:
+            fp_.write(msgpack.dumps(ret))
+
+    def call_parallel(self, cdata, low):
+        '''
+        Call the state defined in the given cdata in parallel
+        '''
+        proc = multiprocessing.Process(
```
Comment (s0undt3ch) [inline salt/state.py]: Use salt.utils.process.MultiprocessingProcess to get logging from the spawned process. Or the SignalHandlingMultiprocessingProcess. Whichever fits your needs.

**Comment by ninja-** [comment]
great work @thatch45 !! I am going to test this and get some feedback as soon as possible.

Consider my example:
My total states run time is 10s.
Of that, 3s is the cloudflare task which updates DNS details, and 7s are the 'normal' tasks.

If I set cloudflare task to run in parallel on start, this should clearly give me free 3 seconds and reduce run time to 7s.

I'll check if that's how it's going to work on my setup 😄

File: salt/state.py [hunk 3]
```diff
@@ -1904,6 +1963,10 @@ def call_chunks(self, chunks):
                 if self.check_failhard(low, running):
                     return running
             self.active = set()
+        while True:
+            if self.reconcile_procs(running):
+                break
+            time.sleep(0.2)
```
Comment (ninja-) [inline salt/state.py]: woah. 200ms seems like way too much...

File: salt/state.py [hunk 4]
```diff
@@ -1994,6 +2087,10 @@ def check_requisite(self, low, running, chunks, pre=False):
                 if tag not in run_dict:
                     fun_stats.add('unmet')
                     continue
+                if run_dict[tag].get('proc'):
+                    # Run in parallel, first wait for a touch and then recheck
+                    time.sleep(0.2)
```
Comment (ninja-) [inline salt/state.py]: Again, this magic value seems like a bit too high, especially if this parallel task is going to block other dependent tasks.

**Review by ninja- — CHANGES_REQUESTED** [review]
This should work pretty well, but I am concerned about the 200ms magic value being too high.

File: salt/state.py (hunk 3 above)
Comment (ninja-) [inline salt/state.py]: If we are waiting forever anyway why not use waitpid() ?

Comment (thatch45) [inline salt/state.py]: good call, I changed them to 0.01

Comment (ninja-) [inline salt/state.py]: wouldn't waitpid() be better?

**Comment by mirceaulinic** [comment]
Fantastic addition @thatch45! Thanks you very much!

Would someone mind explaining the following to me:

```python
time.sleep(0.01)
```

Why is this needed and why the value of 0.01?

Thanks,
Mircea

**Comment by ninja-** [comment]
@mirceaulinic basically, the tasks are started as seperate processess, and the code checks every 10ms if any of background tasks has finished to unblock executions on tasks dependent of them

**Comment by ninja-** [comment]
@thatch45 I am wondering if we should close parent ZeroMQ connections in the forked process so tasks like file.managed etc. won't break with parallel: true? Is this handled at some point in salt's code?

**Comment by thatch45** [comment]
@ninja- this is a good point, I need to test some of this...

**Comment by thatch45** [comment]
As for using waitpid, I do think that this would be a better solution, but at the same time I think this approach makes the code more flexible for now. So I will say that we can move to it later but we should get this functional first

**Comment by damon-atkins** [comment]
Need to be careful with waitpid and windows. Might be a case to use windows specific python functions.

File: doc/ref/states/parallel.rst [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+The `sleep 10` will be started first, then the state system will block on
+starting nginx until the `sleep 10` completes. Once nginx has been ensured to
+be running then the `sleep 5` will start.
+
+This means that the order of evaluation of Salt States and requisites are
+still honored, and given that in the above case, `parallel: True` does not
+actually speed things up.
+
+To run the above state much faster make sure that the `sleep 5` is evaluated
+before the nginx this this:
```
Comment (cachedout) [inline doc/ref/states/parallel.rst]: `this this`?

File: salt/state.py [hunk 6]
```diff
@@ -1645,6 +1649,56 @@ def requisite_in(self, high):
...
+                'name': name,
+                'changes': {},
+                'comment': 'An exception occurred in this state: {0}'.format(
+                    trb)
+            }
+        troot = os.path.join(self.opts['cachedir'], self.jid)
+        tfile = os.path.join(troot, tag)
+        if not os.path.isdir(troot):
+            os.makedirs(troot)
+        with open(tfile, 'wb+') as fp_:
```
Comment (cachedout) [inline salt/state.py]: `salt.utils.fopen`?

File: salt/state.py [hunk 7]
```diff
@@ -1920,6 +1983,35 @@ def check_failhard(self, low, running):
...
+                    ret_cache = os.path.join(self.opts['cachedir'], self.jid, tag)
+                    if not os.path.isfile(ret_cache):
+                        ret = {'result': False,
+                               'comment': 'Parallel process failed to return',
+                               'name': running[tag]['name'],
+                               'changes': {}}
+                    try:
+                        with salt.utils.fopen(ret_cache, 'rb') as fp_:
+                            ret = msgpack.loads(fp_.read())
+                    except OSError, IOError:
```
Comment (cachedout) [inline salt/state.py]: `except (OSError, IOError):` for Python 3 compat, please.

**Comment by thatch45** [comment]
ok, I pushed some fixes

**Comment by cachedout** [comment]
Thanks. I played around a bit with the timing on this to make sure it wasn't going to use too much CPU sorting through the process table in a loop like that. Seems totally fine, so this default is a good one. This should be ready after this test run.

**Review by cachedout — APPROVED** [review]

**Comment by ninja-** [comment]
@thatch45 I checked the matter of zeromq connections(whether it has a chance to break if the parallel task is file.managed etc. or the process forks).
<s>So the proper way to avoid the problem is to set FD_CLOEXEC on the file descriptor, just like in salt.utils.fopen(would be nice to have a Windows fallback but ehh...)
Unless I missed something, this flag is not set on ZeroMQ socket.
It can be done like on the snippet here: http://stackoverflow.com/a/27910874</s>

EDIT: FD_CLOEXEC probably wouldn't trigger because Salt is using fork(..) without exec(...). Hm.
If I am right here, then the part in salt.utils.fopen is also a dead code.

After implementing that I'd feel more safe about going with parallel states or working on SaltCacheLoader knowing that it won't break on threads or fork.

Would you maybe have an idea on how to close the ZeroMQ sockets in child process with minimal patch length?

**Pull request merged**